docs/tutorial/tutorial-10.rst

```diff
@@ -4,17 +4,479 @@
+      Installing icons/helloworld-58.png as 58px application icon... done
+      Installing icons/helloworld-60.png as 60px application icon... done
+      Installing icons/helloworld-76.png as 76px application icon... done
+      Installing icons/helloworld-80.png as 80px application icon... done
+      Installing icons/helloworld-87.png as 87px application icon... done
+      Installing icons/helloworld-120.png as 120px application icon... done
+      Installing icons/helloworld-152.png as 152px application icon... done
+      Installing icons/helloworld-167.png as 167px application icon... done
+      Installing icons/helloworld-180.png as 180px application icon... done
+      Installing icons/helloworld-1024.png as 1024px application icon... done
+
+      [helloworld] Removing unneeded app content...
+      Removing unneeded app bundle content... done
+
+      [helloworld] Application updated.
+
+With our icons installed, we can now run our app with the new icon:
+
+.. tabs::
+
+  .. group-tab:: macOS
+
+    .. code-block:: console
+
+      (beeware-venv) $ briefcase run
+
+      [helloworld] Starting app...
+
+  .. group-tab:: Linux
+
+    .. code-block:: console
+
+      (beeware-venv) $ briefcase run
+
+      [helloworld] Starting app...
+
+  .. group-tab:: Windows
+
+    .. code-block:: doscon
+
+      (beeware-venv) C:\...>briefcase run
+
+      [helloworld] Starting app...
+
+  .. group-tab:: Android
+
+    .. code-block:: console
+
+      (beeware-venv) $ briefcase run android
+
+      [helloworld] Starting app...
+
+  .. group-tab:: iOS
+
+    .. code-block:: console
+
+      (beeware-venv) $ briefcase run iOS
+
+      [helloworld] Starting app...
+
+When you start the app on a desktop platform, you might briefly notice a green
+snake icon... but then the gray bee icon returns. What happened?
+
+Well, an application really has *three* different icons:
+
+* The icon for the app's binary;
+* The icon that the app uses at runtime; and
+* The icon used by the app's installation mechanism.
+
+In most cases, these three icons will be the same - but they don't have to be.
+Briefcase is able to set the first and third of these icons; however, the icon
+used at runtime is determined by the GUI framework. Many GUI frameworks will
+fall back to the icon in the application binary at runtime. However, in some
+cases, this won't be possible; and some frameworks allow you to modify the
+binary to indicate something about the application state (for example, to show
+an unread indicator on the app's icon).
+
+So - in addition to configuring Briefcase to use the new icon for the binary and
+installer, we also need to provide updated icons for our app to use at runtime.
+
+If you look in the app's ``src/helloworld/resources`` folder, you can see the
+icons that are being used by Toga at runtime. We can overwrite these files with
+versions from our ``icons`` folder, then update and re-run our app. We use the
```

Comment:
Having different icons for the app executable and the running app process seems too advanced a topic for a tutorial. And since very few apps will actually want them to be different, leaving the user with two identical icon directories to maintain is giving them extra work for no benefit.

So I suggest keeping the tutorial focused on a single icon set in the default location, and explaining this binary/runtime distinction in the Briefcase icon documentation instead.

Reply:
I agree this is a complex topic to be covering in an introductory tutorial; however, on the basis that something that is hard to explain probably isn't a good design, I think this is evidence that the underlying feature of app icons has issues.

This distinction between "runtime" and "app binary" icons really only matters because Toga is *explicitly* setting a runtime icon. You *can* provide an icon in the construction of `toga.App`; but almost all apps (including the tutorial) fall back onto the default behavior of loading `resources/app_name` as the default icon. 

The app template for each platform *also* includes an icon. This is what gets installed when you do an `update --update-resources`. The current `briefcase-template` setup is essentially using `src/appname/resources/appname` as an *input*, so that it can generate an icon in the app bundle *and* have the same icon available at runtime.

However, this poses 2 problems:
1. As the original icons are in the `src/appname` folder, they are packaged as app resources - which means that a  bundled Windows app will include a macOS `.icns` and GTK `.png` file
2. It requires duplication of the icon file so it can be found both as a runtime resource and by the app binary

If the app is also published as a mobile app, the situation gets worse because of the number of icons needed to support Android and iOS.

If you *don't* explicitly specify an icon at runtime, the icon that is used is the default for the app binary - which is what we're specifying with Briefcase. So, if we modify Toga to *not* explicitly set an icon, and fall back to the default icon for the binary, we don't need to explicitly include the icons as runtime resources. 

This then leaves the case of a Toga app that isn't using Briefcase at all (i.e., `python -m myapp`). A Briefcase app running in dev mode is functionally equivalent to this.

So - I'd like to suggest modifying Toga's app icon behavior:
* If an app explicitly specifies an icon (either by setting `app.icon`, or passing in an icon when constructing the `toga.App` instance) that icon is used.
* If there's no explicit icon, and `sys.executable` is `python`, `python3` or `python3.X`, Toga falls back to the current `resources/app_name` behaviour. 
* If there's no explicit icon and `sys.executable` is anything else, it falls back to the binary's default icon.

This would allow us to modify the default app template to remove the `icon` definition *and* the default icon images in `resources`, and it would avoid shipping macOS icons with Windows apps. The tutorial can then stop after the successful `update --update-resources` pass. It also retains basic Toga branding for the `python -m` case (and, by extension, the `briefcase dev` case).

One additional change could be made to improve the ergonomics of the `briefcase dev` case: Modify Briefcase to include a `BRIEFCASE_APP_ICON` environment variable in the runtime environment if the app configuration specifies a value for `icon`. This would only exist in the dev environment, and would precedence over the 2 fallback options.

I'm not 100% sold on this last change. I'm not all that happy about introducing an explicit dependency between Briefcase and Toga, especially when it's purely for dev mode; but I'm not sure if it's easier to explain that "dev mode doesn't have the app icon", or just include that in the list of things that is different between dev mode and run mode. 

Thoughts?

Reply:
To be clear, the sys.executable variant is as much for Toga branding purposes, rather than run/dev differentiation. The default icons for running Python apps aren't consistent across platforms (and in some cases, quite bad), and don't give any indicator that the app is a GUI (much less Toga) app.

To much lesser extent, it's also a backwards compatibility detail; if you've got a Toga script today with a custom icon, the `resources/app_name` icon will be used. Admittedly, this is a less common user case, and I'm not fundamentally opposed to this being a backwards incompatibility - but given that we get a branding boost as well, I figured we might as well keep it.

Reply:
Where does the "branding boost" come from? Is it an implicit fourth bullet which says "if none of the above succeed, use the Toga logo"? That seems like an awful lot of "if .. and not" conditions that the user would have to read through to work out what's going on.

The current Toga [icon documentation](https://toga.readthedocs.io/en/stable/reference/api/app.html#toga.App) is much simpler:

> The [icon](https://toga.readthedocs.io/en/stable/reference/api/resources/icons.html#c.IconContent) for the app. If not provided, Toga will attempt to load an icon from resources/app_name, where app_name is defined above. If no resource matching this name can be found, a warning will be printed, and the app will fall back to a default icon.

What if we just change this from "a default icon" to "the icon of the app's executable file", and remove "a warning will be printed"? That would preserve the `resources/app_name` behavior for apps using the previous template, and fall back on the executable's icon for apps using the new template, without making the behavior any more complicated than it already is.

Reply:
The implementation I've got on beeware/toga#2527 clarifies (and documents) the options; I've also removed some cases when the warning is printed. It's probably easiest to continue the discussion there where it's tied to specifics. Once those set of changes are finalised, I'll update the tutorial to match.